Assets/PatchKit Patcher/Scripts/AppData/Local/Pack1Unarchiver.cs

```diff
@@ -68,35 +82,68 @@ public void Unarchive(CancellationToken cancellationToken)
             DebugLogger.Log("Unpacking " + _metaData.Files.Length + " files...");
             foreach (var file in _metaData.Files)
             {
-                OnUnarchiveProgressChanged(file.Name, file.Type == "regular", entry, _metaData.Files.Length, 0.0);
+                OnUnarchiveProgressChanged(file.Name, file.Type == Pack1Meta.regularFileType, entry, _metaData.Files.Length, 0.0);
 
                 var currentFile = file;
                 var currentEntry = entry;
-                Unpack(file, progress =>
+
+                if (CanUnpack(file))
                 {
-                    OnUnarchiveProgressChanged(currentFile.Name, currentFile.Type == "regular", currentEntry, _metaData.Files.Length, progress);
-                });
+                    Unpack(file, progress =>
+                    {
+                        OnUnarchiveProgressChanged(currentFile.Name, currentFile.Type == Pack1Meta.regularFileType, currentEntry, _metaData.Files.Length, progress);
+                    });
+                }
 
-                OnUnarchiveProgressChanged(file.Name, file.Type == "regular", entry, _metaData.Files.Length, 1.0);
+                OnUnarchiveProgressChanged(file.Name, file.Type == Pack1Meta.regularFileType, entry, _metaData.Files.Length, 1.0);
 
                 entry++;
             }
             DebugLogger.Log("Unpacking finished succesfully!");
         }
 
+        public void UnarchiveSingleFile(Pack1Meta.FileEntry file, CancellationToken cancellationToken)
+        {
+            OnUnarchiveProgressChanged(file.Name, file.Type == Pack1Meta.regularFileType, 0, 1, 0.0);
+
+            if (!CanUnpack(file))
+            {
+                throw new ArgumentOutOfRangeException("file", file, null);
+            }
+
+            Unpack(file, progress => OnUnarchiveProgressChanged(file.Name, file.Type == Pack1Meta.regularFileType, 1, 1, progress));
+
+            OnUnarchiveProgressChanged(file.Name, file.Type == Pack1Meta.regularFileType, 0, 1, 1.0);
+        }
+
+        private bool CanUnpack(Pack1Meta.FileEntry file)
+        {
+            if (file.Type != Pack1Meta.regularFileType)
+            {
+                return true;
+            }
+
+            if (_range.Start == 0 && _range.End == -1)
+            {
+                return true;
+            }
+
+            return file.Offset > _range.Start && file.Offset + file.Size < _range.End;
```

Comment:
Are you sure that it shouldn't be `file.Offset >= _range.Start`? Also _range.End is inclusive, so `file.Offset + file.Size <= _range.End` (most probably?).

To avoid misakes like that it's better do to something like:
`return _range.Includes(new BytesRange(file.Offset, file.Offset + file.Size - 1))`

or even
`return _range.Includes(file.toBytesRange())`

Reply:
@Dzejkop if you decide to create `Includes` method please do it by either creating extension method for `BytesRange` or as `private static` method inside `Pack1Unarchiver`.